packages/contracts-periphery/contracts/universal/op-nft/OptimistInviter.sol

```diff
@@ -0,0 +1,228 @@
+                _accounts[i],
+                bytes32("optimist.can-invite"),
+                abi.encode(_inviteCount)
+            );
+
+            unchecked {
+                i++;
+            }
+        }
+    }
+
+    /**
+     * @notice Allows anyone to commit a received signature along with the address to claim to.
+     *         This is necessary to prevent front-running when the invitee is claiming the invite.
+     *
+     * @param _commitment A hash of the claimer and signature concatenated.
+                          keccak256(abi.encode(_claimer, _signature))
+     */
+    function commitInvite(bytes32 _commitment) public {
+        commitments[_commitment] = true;
+    }
+
+    /**
+     * @notice Allows anyone to reveal a commitment and claim an invite.
+     *         The claimer ++ signature pair should have been previously committed using
+     *         commitInvite. Doesn't require that the claimer is calling this function.
+     *
+     * @param _claimer Address that will be granted the invite.
+     * @param _claimableInvite ClaimableInvite struct containing the issuer and nonce.
+     * @param _signature Signature signed over the claimable invite.
+     */
+    function claimInvite(
+        address _claimer,
+        ClaimableInvite calldata _claimableInvite,
+        bytes memory _signature
+    ) public {
+        // Make sure the claimer and signature have been committed.
+        require(
+            commitments[keccak256(abi.encode(_claimer, _signature))],
+            "OptimistInviter: claimer and signature have not been committed yet"
+        );
+
+        // Generate a EIP712 typed data hash to compare against the signature.
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    CLAIMABLE_INVITE_TYPEHASH,
+                    _claimableInvite.issuer,
+                    _claimableInvite.nonce
+                )
+            )
+        );
+
+        // Uses SignatureChecker, which supports both regular ECDSA signatures from EOAs as well as
+        // ERC-1271 signatures from contract wallets or multi-sigs. This means that if the issuer
+        // wants to revoke a signature, they can use a smart contract wallet to issue the signature,
+        // then invalidate the signature after issuing it.
+        require(
+            SignatureCheckerUpgradeable.isValidSignatureNow(
+                _claimableInvite.issuer,
+                digest,
+                _signature
+            ),
+            "OptimistInviter: invalid signature"
+        );
+
+        // The issuer includes a pseudorandom nonce in the signature to prevent replay attacks.
+        // This checks that the nonce has not been used for this issuer before. The nonces are
+        // scoped to the issuer address, so the same nonce can be used by different issuers without
+        // clashing.
+        require(
+            !usedNonces[_claimableInvite.issuer][_claimableInvite.nonce],
```

Comment:
slit style nit: i find `== false` easier to read than `! ...`
the less cog overhead for reviewers when writing smart contracts, the better